x/auth/ante/unordered.go

```diff
@@ -26,49 +45,195 @@ var _ sdk.AnteDecorator = (*UnorderedTxDecorator)(nil)
 	// maxUnOrderedTTL defines the maximum TTL a transaction can define.
-	maxUnOrderedTTL uint64
-	txManager       *unorderedtx.Manager
+	maxTimeoutDuration time.Duration
+	txManager          *unorderedtx.Manager
+	sha256Cost         uint64
 }
 
-func NewUnorderedTxDecorator(maxTTL uint64, m *unorderedtx.Manager) *UnorderedTxDecorator {
+func NewUnorderedTxDecorator(
+	maxDuration time.Duration,
+	m *unorderedtx.Manager,
+	gasCost uint64,
+) *UnorderedTxDecorator {
 	return &UnorderedTxDecorator{
-		maxUnOrderedTTL: maxTTL,
-		txManager:       m,
+		maxTimeoutDuration: maxDuration,
+		txManager:          m,
+		sha256Cost:         gasCost,
+	}
+}
+
+func (d *UnorderedTxDecorator) AnteHandle(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	_ bool,
+	next sdk.AnteHandler,
+) (sdk.Context, error) {
+	if err := d.ValidateTx(ctx, tx); err != nil {
+		return ctx, err
 	}
+	return next(ctx, tx, false)
 }
 
-func (d *UnorderedTxDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+func (d *UnorderedTxDecorator) ValidateTx(ctx context.Context, tx sdk.Tx) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sdkTx, ok := tx.(sdk.Tx)
+	if !ok {
+		return fmt.Errorf("invalid tx type %T, expected sdk.Tx", tx)
+	}
+
 	unorderedTx, ok := tx.(sdk.TxWithUnordered)
 	if !ok || !unorderedTx.GetUnordered() {
 		// If the transaction does not implement unordered capabilities or has the
 		// unordered value as false, we bypass.
-		return next(ctx, tx, simulate)
+		return nil
 	}
 
-	// TTL is defined as a specific block height at which this tx is no longer valid
-	ttl := unorderedTx.GetTimeoutHeight()
-
-	if ttl == 0 {
-		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "unordered transaction must have timeout_height set")
+	headerInfo := sdkCtx.HeaderInfo()
+	timeoutTimestamp := unorderedTx.GetTimeoutTimeStamp()
+	if timeoutTimestamp.IsZero() || timeoutTimestamp.Unix() == 0 {
+		return errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"unordered transaction must have timeout_timestamp set",
+		)
 	}
-	if ttl < uint64(ctx.BlockHeight()) {
-		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "unordered transaction has a timeout_height that has already passed")
+	if timeoutTimestamp.Before(headerInfo.Time) {
+		return errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"unordered transaction has a timeout_timestamp that has already passed",
+		)
 	}
-	if ttl > uint64(ctx.BlockHeight())+d.maxUnOrderedTTL {
-		return ctx, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "unordered tx ttl exceeds %d", d.maxUnOrderedTTL)
+	if timeoutTimestamp.After(headerInfo.Time.Add(d.maxTimeoutDuration)) {
+		return errorsmod.Wrapf(
+			sdkerrors.ErrInvalidRequest,
+			"unordered tx ttl exceeds %s",
+			d.maxTimeoutDuration.String(),
+		)
 	}
 
-	txHash := sha256.Sum256(ctx.TxBytes())
+	// consume gas in all exec modes to avoid gas estimation discrepancies
+	sdkCtx.GasMeter().ConsumeGas(d.sha256Cost, "consume gas for calculating tx hash")
```

Comment:
One thing I want us to follow up on is "why is the default cost what it is?"

Reply:
https://github.com/cosmos/cosmos-sdk/issues/23816